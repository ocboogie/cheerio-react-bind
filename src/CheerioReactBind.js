import React from "react";
import PropTypes from "prop-types";

export default class CheerioReactBind extends React.Component {
  constructor(props) {
    super(props);
    this.props.$elem.data("update", this.update.bind(this));
    this.update(true);
  }

  update(first = false) {
    const children = [];
    const $children = this.props.$elem.children();
    if ($children.length > 0) {
      this.props.$elem.children().each((index, elem) => {
        const $child = this.props.$(elem);
        children.push(
          <CheerioReactBind
            errorHandler={this.props.errorHandler}
            key={index}
            $={this.props.$}
            $elem={$child}
            tags={this.props.tags}
            tagRenderer={this.props.tagRenderer}
            location={`${
              this.props.location === "/" ? "" : this.props.location
            }/${$child.index()}:${$child.prop("tagName").toLowerCase()}`}
          />
        );
      });
    } else {
      children.push(this.props.$elem.text());
    }
    const state = {
      attributes: this.props.$elem.attr(),
      children
    };
    if (first) {
      this.state = state;
    } else {
      this.setState(state);
    }
  }

  render() {
    const { $elem } = this.props;
    const tagName = $elem.prop("tagName").toLowerCase();
    if (!this.props.tagRenderer && !this.props.tags) {
      throw new TypeError(
        'You must pass a "tagRenderer" prop or a "tags" prop.'
      );
    }

    if (this.props.tagRenderer) {
      const TagHander = this.props.tagRenderer;
      return (
        <TagHander
          location={this.props.location}
          tagName={tagName}
          attributes={this.state.attributes}
        >
          {this.state.children}
        </TagHander>
      );
    }

    if (!Object.prototype.hasOwnProperty.call(this.props.tags, tagName)) {
      const errMsg = `Unknown tag "${tagName}".`;
      if (this.props.errorHandler) {
        this.props.errorHandler(errMsg);
        return <div />;
      }
      throw new TypeError(errMsg);
    }
    const Component = this.props.tags[tagName];
    return (
      <Component {...this.state.attributes}>{this.state.children}</Component>
    );
  }
}

/* eslint-disable react/require-default-props, consistent-return */
CheerioReactBind.propTypes = {
  $: (props, propName) => {
    const $ = props[propName];
    if (!$) {
      return new Error('Prop type "$" is required for CheerioReactBind');
    }
    if (!$.root) {
      return new Error(
        'Prop type "$" must be a valid Cheerio object for CheerioReactBind'
      );
    }
  },
  $elem: (props, propName) => {
    const $elem = props[propName];
    if (!$elem) {
      return new Error('Prop type "$elem" is required for CheerioReactBind');
    }
    if (!$elem.data || !$elem.prop) {
      return new Error(
        'Prop type "$" must be a valid Cheerio element for CheerioReactBind'
      );
    }
  },
  // eslint-disable-next-line react/forbid-prop-types
  tags: PropTypes.object,
  tagRenderer: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
  errorHandler: PropTypes.func,
  location: PropTypes.string
};
/* eslint-enable react/require-default-props, consistent-return */

CheerioReactBind.defaultProps = {
  location: "/"
};
