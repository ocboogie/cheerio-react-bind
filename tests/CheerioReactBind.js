/* eslint react/prop-types: "off" */
import React from "react";
import Cheerio from "cheerio";
import * as Enzyme from "enzyme";

import CheerioReactBind, { update } from "../src/index";

const tags = {
  div: ({ children }) => <div>{children}</div>,
  h1: ({ children }) => <h1>{children}</h1>,
  p: ({ children }) => <p>{children}</p>
};

const $ = Cheerio.load(
  `
  <div foo=bar>
    <p>Paragraph</p>
  </div>
`,
  { xmlMode: true }
);

test("modifying the Cheerio dom updates the React dom with custom tags", () => {
  const wrapper = Enzyme.mount(
    <CheerioReactBind tags={tags} $elem={$("div").first()} $={$} />
  );
  expect(wrapper.html()).toMatchSnapshot();

  $("div").append(`<div>test</div>`);
  update($("div"));
  expect(wrapper.html()).toMatchSnapshot();

  $("p").text("Paragraph test");
  update($("p"));
  expect(wrapper.html()).toMatchSnapshot();
});

describe("error handling", () => {
  test("throwns when there is a unknown tag", () => {
    Enzyme.mount(
      <CheerioReactBind tags={tags} $elem={$("div").first()} $={$} />
    );
    $("div").append("<notag />");
    expect(() => {
      update($("div"));
    }).toThrow(new TypeError('Unknown tag "notag".'));
  });

  test("errorHandling runs when an error is thrown", () => {
    const mockErrorHandling = jest.fn();
    Enzyme.render(
      <CheerioReactBind
        errorHandler={mockErrorHandling}
        tags={tags}
        $elem={$("div").first()}
        $={$}
      />
    );
    $("div").append("<notag />");
    update($("div"));

    expect(mockErrorHandling.mock.calls[0][0]).toBe('Unknown tag "notag".');
  });
});

describe("tagRenderer", () => {
  test("passed children, attributes and tagName", () => {
    const $mock = Cheerio.load(
      `
    <foo bar="baz"><qux /></foo>
  `,
      { xmlMode: true }
    );
    const tagRenderer = jest.fn(() => <div />);
    Enzyme.mount(
      <CheerioReactBind
        tagRenderer={tagRenderer}
        $elem={$mock("foo").first()}
        $={$mock}
      />
    );
    expect(tagRenderer.mock.calls[0][0]).toMatchObject({
      tagName: "foo",
      attributes: { bar: "baz" }
    });
    expect(() =>
      Enzyme.shallow(tagRenderer.mock.calls[0][0].children[0])
    ).not.toThrow();
  });

  test("renders", () => {
    const $mock = Cheerio.load(
      `
    <foo bar="baz"><qux>quux</qux></foo>
  `,
      { xmlMode: true }
    );
    const tagRenderer = ({ tagName, children }) => {
      if (tagName === "foo") {
        return <span>{children}</span>;
      } else if (tagName === "qux") {
        return <h2>{children}</h2>;
      }
      return null;
    };
    const wrapper = Enzyme.mount(
      <CheerioReactBind
        tagRenderer={tagRenderer}
        $elem={$mock("foo").first()}
        $={$mock}
      />
    );
    expect(wrapper.html()).toMatchSnapshot();
  });
});

test("throwns when then neither tags nor tagRenderer", () => {
  expect(() => {
    Enzyme.render(<CheerioReactBind $elem={$("div").first()} $={$} />);
  }).toThrow(
    new TypeError('You must pass a "tagRenderer" prop or a "tags" prop.')
  );
});

test("tagRenderer gets passed the location of the tag", () => {
  const $mock = Cheerio.load(
    `
    <div>
      <foo></foo>
      <bar><baz></baz></bar>
    </div>
`,
    { xmlMode: true }
  );
  const tagRenderer = ({ tagName, location, children }) => {
    if (tagName === "baz") {
      expect(location).toBe("/1:bar/0:baz");
    }
    return <div>{children}</div>;
  };
  Enzyme.mount(
    <CheerioReactBind
      tagRenderer={tagRenderer}
      $elem={$mock("div").first()}
      $={$mock}
    />
  );
});
