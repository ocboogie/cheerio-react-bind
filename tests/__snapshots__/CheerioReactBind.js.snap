// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tagRenderer renders 1`] = `"<span><h2>quux</h2></span>"`;

exports[`update the Cheerio dom when the React dom is updated 1`] = `"<div><p>Paragraph</p></div>"`;

exports[`update the Cheerio dom when the React dom is updated 2`] = `"<div><p>Paragraph</p><div>test</div></div>"`;

exports[`update the Cheerio dom when the React dom is updated 3`] = `"<div><p>Paragraph test</p><div>test</div></div>"`;
